### PR TITLE
Cap timer duration

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
Restrict timer duration to max 100 years
* Workflow
  * Backoff
  * RunTimeout
  * ExecutionTimeout
* ActivityTask
  * ScheduleToCloseTimeout
  * ScheduleToStartTimeout
  * StartToCloseTimeout
  * HeartbeatTimeout
* WorkflowTask
  * StartToCloseTimeout 
* UserTimer

**Why?**

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

**Comment by yycptt** [comment]
My only concern is that this could break existing oss users. The existing behavior when time.Time is valid but unixNano overflows is that a garbage task will be created and never delete. But user workflow can still run and won't see the timer firing. 

An alternative solution is just drop those timers or commands.

File: service/frontend/workflow_handler.go [hunk 1]
```diff
@@ -4691,15 +4692,15 @@ func (wh *WorkflowHandler) validateRetryPolicy(namespaceName namespace.Name, ret
...
 	}
 
-	if timestamp.DurationValue(request.GetWorkflowRunTimeout()) < 0 {
+	if !timer.ValidateTimer(request.GetWorkflowRunTimeout()) {
 		return errInvalidWorkflowRunTimeoutSeconds
 	}
 
-	if timestamp.DurationValue(request.GetWorkflowTaskTimeout()) < 0 {
+	if !timer.ValidateTimer(request.GetWorkflowTaskTimeout()) {
 		return errInvalidWorkflowTaskTimeoutSeconds
```
Comment (mastermanu) [inline service/frontend/workflow_handler.go:4704]: should the error message return more information indicating what the min/maximum is?

**Comment by wxing1292** [comment]
> The existing behavior when time.Time is valid but unixNano overflows is that a garbage task will be created and never delete.

the timer which overflow will never fire, which is still a bug.

**Comment by wxing1292** [comment]
> An alternative solution is just drop those timers or commands.

which is equivalent to intentionally lose data

**Comment by wxing1292** [comment]
i will let OSS team to take over this PR

cc @yiminc @yycptt

**Comment by Neirpyc** [comment]
Why are you setting an arbitrary duration (100 years) instead of checking that timer < unixNanoMax - now() ?

**Comment by yycptt** [comment]
> Why are you setting an arbitrary duration (100 years) instead of checking that timer < unixNanoMax - now() ?

Using now() won't work as downstream will calculate the target timestamp again using now() + the timer duration in request, so there's still a chance of exceeding the limit. 

We will cap the timer duration instead of returning error when the limit is exceeded. So to workflow there's no difference. Well unless the workflow & temporal cluster will run for more than 100 years.

**Review by yiminc — APPROVED** [review]

**Pull request merged**